util/topgen/merge.py

```diff
@@ -477,6 +512,77 @@ def no_device_in_range(xbar, f, t):
     return True
 
 
+def get_next_base_addr(addr, xbar, name):
+    """Return the least value of base_addr of the IP greater than addr
+
+    """
+    if isinstance(addr, str):
+        value = int(addr, 0)
+    else:
+        assert isinstance(addr, int)
+        value = addr
+
+    device_list = [
+        x for x in xbar["nodes"]
+        if x["type"] == "device" and not x["name"] == name
+    ]
+
+    try:
+        addrs = [a for r in device_list for a in r["addr_range"]]
+    except KeyError:
+        log.error("Address range is wrong.\n {}".format(
+            [x for x in device_list if not "addr_range" in x]))
+        raise SystemExit()
+
+    sorted_list = sorted(addrs, key=lambda k: int(k["base_addr"], 0))
+
+    gte_list = [x for x in sorted_list if int(x["base_addr"], 0) > value]
+
+    if len(gte_list) == 0:
+        return -1
+
+    return gte_list[0]
+
+
+def find_pow2_size(addr, next_value):
+    """Find greatest power of 2 value by given addr.
+
+    For instance, (0x4000_0000, 0x21000) and `next_value` as 0x40080000,
+    the result will be 0x8_0000
+
+    But it should return result not exceeding the base_addr's leading one bit
+    position. For instance, if the base_addr is 0x4003_0000, the return value
+    should be less than or equal to 0x2_0000. Cannot be 0x8_0000.
+    """
+    base_addr = int(addr["base_addr"], 0)
+
+    diff = next_value - base_addr
+
+    # Find the least one bit position.
+    # If base_addr is 0, then any value can be used
+    if not base_addr == 0:
+        leading_one = 1
+        while True:
+            if base_addr & leading_one != 0:
+                break
+            leading_one = leading_one << 1
+
+        if leading_one <= diff:
+            diff = leading_one
```

Comment:
@msfschaffner @tjaychen Added this code to limit the maximum power of 2 finding which should be less than or equal to the leading one bit position.

Reply:
hmmm...what's the reason we want to limit this? Sorry if i missed something.  If the size of the block is actually 0x3_0000 for some reason, wouldn't you actually want to align to 0x8_0000 instead of relying on `calc_size`?

Reply:
The reason to find the maximum value of power of 2 is to change the `>=` and `<=` comparisons of the given address into the `OFFSET` and `MASK`. So, the `MASK` value should be smaller than the leading one position.

For instance, if base addr is 0x4003_0000 and the size is 0x1000, then the maximum power of 2 size should be smaller than or equal to 0x3_0000 which results in 0x2_0000. Then the `MASK` value will be `0x1_FFFF` then the address checker logic inside the crossbar can be in this form below:

```systemverilog
addr & (~MASK) == OFFSET
```

Reply:
hmmm...i guess this would imply then for modules starting at 0xN000_0000, it would need a much bigger mask than blocks that follow?

It kind of feels like if we want to do this, maybe it would be better to put down uniform rules on how much space each device can have, and just always allocate that space to it.  For example, each can always have 256KB or something of peripheral space.